src/platform/Tizen/AppPreference.cpp

```diff
@@ -19,145 +19,114 @@
-    VerifyOrExit(preErr == PREFERENCE_ERROR_NONE, err = CHIP_ERROR_INCORRECT_STATE);
-    encodedDataSize = strlen(encodedData);
+    // Make sure that string allocated by preference_get_string() will be freed
+    std::unique_ptr<char, decltype(&::free)> _{ encodedData, &::free };
+    size_t encodedDataSize = strlen(encodedData);
 
+    size_t encodedDataPaddingSize = 0;
     if ((encodedDataSize > 0) && (encodedData[encodedDataSize - 1] == '='))
     {
         encodedDataPaddingSize++;
         if ((encodedDataSize > 1) && (encodedData[encodedDataSize - 2] == '='))
             encodedDataPaddingSize++;
     }
-    expectedDecodedSize = ((encodedDataSize - encodedDataPaddingSize) * 3) / 4;
+    size_t expectedDecodedSize = ((encodedDataSize - encodedDataPaddingSize) * 3) / 4;
 
-    decodedData = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(expectedDecodedSize));
-    VerifyOrExit(decodedData != NULL, err = CHIP_ERROR_INCORRECT_STATE);
+    Platform::ScopedMemoryBuffer<uint8_t> decodedData;
+    VerifyOrReturnError(decodedData.Alloc(expectedDecodedSize), CHIP_ERROR_NO_MEMORY);
 
-    decodedDataSize = Base64Decode(encodedData, static_cast<uint16_t>(encodedDataSize), decodedData);
+    size_t decodedDataSize = Base64Decode(encodedData, static_cast<uint16_t>(encodedDataSize), decodedData.Get());
+    VerifyOrReturnError(dataSize >= decodedDataSize - offset, CHIP_ERROR_BUFFER_TOO_SMALL);
 
-    copy_size = min(dataSize, decodedDataSize - offset);
-    if (getDataSize != NULL)
+    size_t copySize = std::min(dataSize, decodedDataSize - offset);
+    if (getDataSize != nullptr)
     {
-        *getDataSize = copy_size;
+        *getDataSize = copySize;
     }
-    memset(data, 0, dataSize);
-    memcpy(data, decodedData + offset, copy_size);
-
-    ChipLogProgress(DeviceLayer, "Get data [%s:%s]", key, (char *) data);
+    ::memcpy(data, decodedData.Get() + offset, copySize);
 
-    chip::Platform::MemoryFree(decodedData);
-
-    VerifyOrExit(dataSize >= decodedDataSize - offset, err = CHIP_ERROR_BUFFER_TOO_SMALL);
-
-exit:
-    free(encodedData);
-    return err;
+    ChipLogProgress(DeviceLayer, "Get data [%s:%.*s]", key, static_cast<int>(copySize), static_cast<char *>(data));
+    return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR SaveData(const char * key, const uint8_t * data, size_t dataSize)
+CHIP_ERROR SaveData(const char * key, const void * data, size_t dataSize)
 {
-    CHIP_ERROR err             = CHIP_NO_ERROR;
-    int preErr                 = PREFERENCE_ERROR_NONE;
-    char * encodedData         = NULL;
-    size_t encodedDataSize     = 0;
-    size_t expectedEncodedSize = ((dataSize + 3) * 4) / 3;
-
-    err = __IsKeyExist(key);
-    if (err == CHIP_NO_ERROR)
-    {
-        VerifyOrExit(RemoveData(key) == CHIP_NO_ERROR, err = CHIP_ERROR_INCORRECT_STATE);
-    }
-    err = CHIP_NO_ERROR;
+    // Expected size for null-terminated base64 string
+    size_t expectedEncodedSize = (((dataSize + 3 - 1) / 3) * 4) + 1;
```

Comment:
Could we use the BASE64_MAX_DECODED_LEN and BASE64_ENCODED_LEN macros here from Base64.h?

The calculation here seems a bit different than those. Why is that?

Reply:
OK, I will use the macro (I was not aware that such macro exists).

> The calculation here seems a bit different than those. Why is that?
```c++
#define BASE64_ENCODED_LEN(LEN) ((((LEN) + 2) / 3) * 4)
```
The calculation is not different, it's "exactly" the same: `LEN + 2` == `LEN + (3 - 1))`  (I wanted to stress out that we are doing round up here). The lenght calculated here shall include end zero-byte for string termination, hence, `+ 1` at the end :)